helix-core/src/main/java/org/apache/helix/model/InstanceConfig.java

```diff
@@ -656,6 +656,17 @@ public InstanceOperation getInstanceOperation() {
           .build();
     }
 
+    // if instance operation is DISABLE, we override it to ENABLE if HELIX_ENABLED set to true
+    if (activeInstanceOperation.getOperation() == InstanceConstants.InstanceOperation.DISABLE) {
+      // it is not likely that HELIX_ENABLED is unset, because when we set operation to disable,
+      // we always set HELIX_ENABLED to false
+      // If instance is enabled by old version helix (not having instance operation), the instance config
+      // will have HELIX_ENABLED set to true. In this case, we should override the instance operation to ENABLE
+      if ("true".equals(_record.getSimpleField(InstanceConfigProperty.HELIX_ENABLED.name()))) {
+        return new InstanceOperation.Builder().setOperation(InstanceConstants.InstanceOperation.ENABLE).build();
```

Comment:
Shall we write it back?

Reply:
good point.

Reply:
I think it might be dangerous to write back as client may update the config while we write back. It could overwrite the current value.  (Unlike the batch enable/disable one as that field is deprecated and we just need to clean it out)

Since the getter will use the `HELIX_ENABLED` and is compatible with both new/old client before instance operation, I think it is safe to keep it the current way.